modules/Attendance/Services/AttendanceService.php

```
@@ -0,0 +1,920 @@
+            
+        // Get total user count for pagination
+        $totalUserCount = $query->count();
+        $totalItems = $totalUserCount * $dateCount;
+        
+        // Calculate which users we need for the current page
+        $usersPerPage = max(1, ceil($perPage / $dateCount));
+        $userOffset = floor($offset / $dateCount);
+        
+        // Only fetch users needed for this page
+        $allRelevantUsers = $query->with(['professionalData'])
+            ->skip($userOffset)
+            ->take($usersPerPage + 1) // Fetch one extra to ensure we have enough
+            ->get();
+
+        $allRelevantUserIds = $allRelevantUsers->pluck('id')->toArray();
+
+        // Optimization 2: Only select fields we need
+        $realAttendanceRecords = Attendance::query()
+            ->filter($filters)
+            ->select(['id', 'user_id', 'company_id', 'status', 'is_late', 'is_absent', 
+                     'is_holiday', 'day_status', 'clock_in_time', 'clock_out_time', 
+                     'start_time', 'overtime_hours'])
+            ->whereIn('user_id', $allRelevantUserIds)
+            ->whereBetween('start_time', [$startDate, $endDate])
+            ->orderBy('start_time')
+            ->get();
+
+        // Optimization 3: More efficient data structuring
+        $attendanceByUserAndDate = [];
+        foreach ($realAttendanceRecords as $record) {
+            $userId = (string)$record->user_id;
+            // Ensure clock_in_time is a Carbon object before calling methods on it
+            if ($record->clock_in_time instanceof Carbon) {
+                $date = $record->clock_in_time->copy()->setTimezone($timezone)->format('Y-m-d');
+            } else if (is_string($record->clock_in_time)) {
+                $date = Carbon::parse($record->clock_in_time, $timezone)->format('Y-m-d');
+            } else {
+                // Default fallback if clock_in_time is null or invalid
+                $date = Carbon::now($timezone)->format('Y-m-d');
```

Using `Carbon::now()` as fallback might mask data issues. Consider throwing an exception or logging a warning when `clock_in_time` is invalid/null instead of silently defaulting to current time.